File: templates/csharp/BdappsClient.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Options;

namespace bdapps;

/// <summary>
/// bdapps API client — C# port of templates/typescript/bdapps-client.ts.
///
/// <para>One <c>PostAsync</c> helper injects credentials, applies a timeout, and turns
/// non-S1000 responses into typed errors. Every service is a thin wrapper that resolves its
/// endpoint through <see cref="BdappsOptions.RequireEndpoint"/> — so calling an API your
/// application was not provisioned for fails locally with a clear message, rather than as
/// E1309 from the platform.</para>
///
/// <para>Register with <c>IHttpClientFactory</c>:
/// <c>builder.Services.AddHttpClient&lt;BdappsClient&gt;();</c></para>
///
/// <para>SERVER-SIDE ONLY.</para>
/// </summary>
public sealed class BdappsClient
{
    /// <summary>A single outbound call should never hang. Protocol constant, not config.</summary>
    public static readonly TimeSpan Timeout = TimeSpan.FromSeconds(15);

    /// <summary>Platform-side. Worth retrying with backoff.</summary>
    public static readonly IReadOnlySet<string> Transient = new HashSet<string>
    {
        "E1316", "E1318", "E1319", "E1332", "E1341",
        "E1360", "E1363", "E1364", "E1600", "E1601", "E1602", "E1603",
    };

    /// <summary>Provisioning or credentials are wrong. Retrying will never help.</summary>
    public static readonly IReadOnlySet<string> Configuration = new HashSet<string>
    {
        "E1301", "E1302", "E1303", "E1304", "E1305", "E1306", "E1307",
        "E1309", "E1310", "E1311", "E1313", "E1315", "E1322", "E1323",
        "E1324", "E1327", "E1328", "E1329", "E1336", "E1371", "E1381",
        "E1383", "E1387",
    };

    // Codes meaning "the state you wanted already holds".
    public const string BenignRegister = "E1351";   // user already registered
    public const string BenignUnregister = "E1356"; // user not registered
    public const string BenignDebit = "E1379";      // transaction already completed

    // Mandatory on every CaaS call, with exactly one accepted value. It only ever varies by
    // being forgotten, so it is a constant rather than a caller argument.
    public const string PaymentInstrument = "MobileAccount";

    public const string BroadcastConfirmation = "I_HAVE_VERIFIED_THIS_GOES_TO_ALL_SUBSCRIBERS";

    private readonly BdappsOptions _options;
    private readonly HttpClient _http;

    /// <summary>
    /// bdapps hosts have served an incomplete certificate chain, which .NET rejects. Do NOT
    /// install a callback that returns true unconditionally — that lets anyone on the path read
    /// the applicationId and password that can charge your subscribers. Supply the intermediate
    /// CA through a SocketsHttpHandler with a custom trust store instead. See
    /// references/09-security-best-practices.md.
    /// </summary>
    public BdappsClient(HttpClient http, IOptions<BdappsOptions> options)
    {
        _http = http;
        _http.Timeout = Timeout;
        _options = options.Value;
    }

    /* ── Helpers ──────────────────────────────────────────────────────────── */

    private static readonly Regex Separators = new(@"[\s()\-]", RegexOptions.Compiled);

    /// <summary>
    /// Normalise a subscriber address. The ONLY place <c>tel:</c> is added. Accepts an
    /// already-prefixed address, a masked hash, +94…, 0094… or a local 07… number.
    /// </summary>
    public static string ToTelAddress(string msisdn)
    {
        var trimmed = (msisdn ?? string.Empty).Trim();
        if (trimmed.Length == 0)
        {
            throw new ArgumentException("[bdapps] Empty subscriber address", nameof(msisdn));
        }

        if (trimmed.StartsWith("tel:", StringComparison.OrdinalIgnoreCase))
        {
            return trimmed;
        }

        var digits = Separators.Replace(trimmed, string.Empty).TrimStart('+');
        if (digits.StartsWith("00", StringComparison.Ordinal))
        {
            digits = digits[2..];
        }

        if (digits.StartsWith("0", StringComparison.Ordinal) && digits.Length == 10)
        {
            digits = "94" + digits[1..];
        }

        return "tel:" + digits;
    }

    /// <summary>Mask a subscriber address for logging. Never log the raw value.</summary>
    public static string MaskAddress(string address)
    {
        var body = address ?? string.Empty;
        if (body.StartsWith("tel:", StringComparison.OrdinalIgnoreCase))
        {
            body = body[4..];
        }

        return body.Length <= 6
            ? "tel:***"
            : $"tel:{body[..3]}{new string('*', body.Length - 6)}{body[^3..]}";
    }

    /// <summary>A unique, persistable idempotency key for a charge. Max 32 characters.</summary>
    public static string GenerateExternalTrxId() => Guid.NewGuid().ToString("N");

    /* ── Core ─────────────────────────────────────────────────────────────── */

    private async Task<JsonElement> PostAsync(
        string service,
        string url,
        IDictionary<string, object?> body,
        CancellationToken cancellationToken,
        params string[] benignCodes)
    {
        var payload = new Dictionary<string, object?>(body)
        {
            ["applicationId"] = _options.ApplicationId,
            ["password"] = _options.Password,
        };

        using var response = await _http
            .PostAsJsonAsync(url, payload, cancellationToken)
            .ConfigureAwait(false);

        // response.EnsureSuccessStatusCode() is deliberately NOT called: bdapps returns
        // HTTP 200 for application-level failures, and the real outcome is statusCode.
        var data = await response.Content
            .ReadFromJsonAsync<JsonElement>(cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        var statusCode = data.TryGetProperty("statusCode", out var code)
            ? code.GetString() ?? string.Empty
            : string.Empty;

        if (statusCode == "S1000" || benignCodes.Contains(statusCode))
        {
            return data;
        }

        var detail = data.TryGetProperty("statusDetail", out var value)
            ? value.GetString() ?? string.Empty
            : string.Empty;

        throw new BdappsException(statusCode, detail, service);
    }

    /* ── SMS ──────────────────────────────────────────────────────────────── */

    /// <summary>Send an MT SMS to one or more subscribers.</summary>
    public Task<JsonElement> SendSmsAsync(
        IEnumerable<string> to,
        string message,
        CancellationToken cancellationToken = default)
    {
        var recipients = to.Select(ToTelAddress).ToList();
        if (recipients.Contains("tel:all"))
        {
            throw new ArgumentException(
                "[bdapps] Use BroadcastSmsAsync for tel:all — broadcasts must be deliberate.",
                nameof(to));
        }

        return PostAsync(
            "sms-send",
            _options.RequireEndpoint("SmsSend"),
            new Dictionary<string, object?>
            {
                ["message"] = message,
                ["destinationAddresses"] = recipients,
            },
            cancellationToken);
    }

    /// <summary>
    /// Send to the ENTIRE subscribed base. Deliberately separate from
    /// <see cref="SendSmsAsync"/> so it can never be reached by accident — check the subscriber
    /// base size first, and put an authorisation check in front of this.
    /// </summary>
    public Task<JsonElement> BroadcastSmsAsync(
        string message,
        string confirmation,
        CancellationToken cancellationToken = default)
    {
        if (confirmation != BroadcastConfirmation)
        {
            throw new ArgumentException(
                "[bdapps] Broadcast confirmation token missing", nameof(confirmation));
        }

        return PostAsync(
            "sms-send",
            _options.RequireEndpoint("SmsSend"),
            new Dictionary<string, object?>
            {
                ["message"] = message,
                ["destinationAddresses"] = new[] { "tel:all" },
            },
            cancellationToken);
    }

    /* ── USSD ─────────────────────────────────────────────────────────────── */

    /// <summary>
    /// Send a USSD screen. <paramref name="sessionId"/> MUST be the one the platform sent you.
    /// Use "mt-fin" for the final screen — anything else leaves the session hanging until the
    /// network times out.
    /// </summary>
    public Task<JsonElement> SendUssdAsync(
        string sessionId,
        string destinationAddress,
        string message,
        string operation,
        CancellationToken cancellationToken = default)
    {
        if (operation is not ("mt-init" or "mt-cont" or "mt-fin"))
        {
            throw new ArgumentException(
                $"[bdapps] Invalid ussdOperation '{operation}'", nameof(operation));
        }

        return PostAsync(
            "ussd-send",
            _options.RequireEndpoint("UssdSend"),
            new Dictionary<string, object?>
            {
                ["message"] = message,
                ["sessionId"] = sessionId,
                ["ussdOperation"] = operation,
                ["destinationAddress"] = ToTelAddress(destinationAddress),
                ["encoding"] = "440",
                ["version"] = "1.0",
            },
            cancellationToken);
    }

    /* ── Subscription ─────────────────────────────────────────────────────── */

    /// <summary>
    /// Opt a subscriber in. Only call this with recorded, explicit consent. E1351 (already
    /// registered) is accepted as success — the desired state holds.
    /// </summary>
    public Task<JsonElement> RegisterAsync(
        string subscriberId, CancellationToken cancellationToken = default) =>
        SubscriptionAsync(
            "subscription-register", subscriberId, "1", BenignRegister, cancellationToken);

    /// <summary>
    /// Opt a subscriber out. E1356 (not registered) is accepted as success — the desired state
    /// holds.
    /// </summary>
    public Task<JsonElement> UnregisterAsync(
        string subscriberId, CancellationToken cancellationToken = default) =>
        SubscriptionAsync(
            "subscription-unregister", subscriberId, "0", BenignUnregister, cancellationToken);

    private Task<JsonElement> SubscriptionAsync(
        string service,
        string subscriberId,
        string action,
        string benign,
        CancellationToken cancellationToken) =>
        PostAsync(
            service,
            _options.RequireEndpoint("SubscriptionSend"),
            new Dictionary<string, object?>
            {
                ["subscriberId"] = ToTelAddress(subscriberId),
                ["action"] = action,
                ["version"] = "1.0",
            },
            cancellationToken,
            benign);

    /// <summary>Check one subscriber's status. For reconciliation, not per-request gating.</summary>
    public Task<JsonElement> GetSubscriptionStatusAsync(
        string subscriberId, CancellationToken cancellationToken = default) =>
        PostAsync(
            "subscription-status",
            _options.RequireEndpoint("SubscriptionStatus"),
            new Dictionary<string, object?> { ["subscriberId"] = ToTelAddress(subscriberId) },
            cancellationToken);

    /// <summary>
    /// Subscriber base size. Needs no subscriber and charges nothing, which also makes it the
    /// best connectivity and credential smoke test. <c>baseSize</c> comes back as a string.
    /// </summary>
    public async Task<long> QueryBaseAsync(CancellationToken cancellationToken = default)
    {
        var data = await PostAsync(
            "subscription-query-base",
            _options.RequireEndpoint("SubscriptionQueryBase"),
            new Dictionary<string, object?>(),
            cancellationToken).ConfigureAwait(false);

        var raw = data.TryGetProperty("baseSize", out var value) ? value.GetString() : "0";
        return long.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out var size)
            ? size
            : 0;
    }

    /* ── OTP ──────────────────────────────────────────────────────────────── */

    /// <summary>
    /// Send an OTP to a plain mobile number. Rate-limit per number AND per IP before calling,
    /// or the app becomes an SMS-bombing tool. Keep the returned referenceNo server-side.
    /// </summary>
    public Task<JsonElement> RequestOtpAsync(
        string subscriberId,
        IDictionary<string, object?> applicationMetaData,
        CancellationToken cancellationToken = default) =>
        PostAsync(
            "otp-request",
            _options.RequireEndpoint("OtpRequest"),
            new Dictionary<string, object?>
            {
                ["subscriberId"] = ToTelAddress(subscriberId),
                ["applicationMetaData"] = applicationMetaData,
            },
            cancellationToken);

    /// <summary>
    /// Verify an OTP. Valid 60 minutes, maximum 3 attempts — enforce those limits on your side
    /// too. The returned subscriberId is the masked identifier to use for every subsequent call.
    /// </summary>
    public Task<JsonElement> VerifyOtpAsync(
        string referenceNo, string otp, CancellationToken cancellationToken = default) =>
        PostAsync(
            "otp-verify",
            _options.RequireEndpoint("OtpVerify"),
            new Dictionary<string, object?> { ["referenceNo"] = referenceNo, ["otp"] = otp },
            cancellationToken);

    /* ── CaaS ─────────────────────────────────────────────────────────────── */

    /// <summary>
    /// Charge a subscriber's mobile account. THIS MOVES REAL MONEY.
    ///
    /// <para><paramref name="externalTrxId"/> is your idempotency key: generate it with
    /// <see cref="GenerateExternalTrxId"/>, PERSIST IT, then call this. There are deliberately
    /// no retries — a timeout does NOT mean the charge failed. Resolve unknown outcomes by
    /// re-calling with the SAME id — the platform de-duplicates on it. E1379
    /// (already completed) is accepted as success.</para>
    ///
    /// <para><paramref name="amount"/> is <c>decimal</c>, never <c>double</c>.</para>
    /// </summary>
    public Task<JsonElement> DebitAsync(
        string subscriberId,
        decimal amount,
        string externalTrxId,
        string currency = "BDT",
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(externalTrxId))
        {
            throw new ArgumentException(
                "[bdapps] externalTrxId is required and must be persisted first",
                nameof(externalTrxId));
        }

        if (externalTrxId.Length > 32)
        {
            throw new ArgumentException(
                "[bdapps] externalTrxId must be 32 characters or fewer", nameof(externalTrxId));
        }

        return PostAsync(
            "caas-direct-debit",
            _options.RequireEndpoint("CaasDebit"),
            new Dictionary<string, object?>
            {
                ["externalTrxId"] = externalTrxId,
                ["subscriberId"] = ToTelAddress(subscriberId),
                ["paymentInstrumentName"] = PaymentInstrument,
                ["amount"] = amount.ToString(CultureInfo.InvariantCulture),
                ["currency"] = currency,
            },
            cancellationToken,
            BenignDebit);
    }

    /// <summary>
    /// Query chargeable balance. Advisory only: the balance can change between this and the
    /// debit. Always handle E1378 on the debit regardless of what this returned.
    /// </summary>
    public Task<JsonElement> QueryBalanceAsync(
        string subscriberId,
        string currency = "BDT",
        CancellationToken cancellationToken = default) =>
        PostAsync(
            "caas-balance-query",
            _options.RequireEndpoint("CaasBalance"),
            new Dictionary<string, object?>
            {
                ["subscriberId"] = ToTelAddress(subscriberId),
                ["paymentInstrumentName"] = PaymentInstrument,
                ["currency"] = currency,
            },
            cancellationToken);

    /* ── Extension point ───────────────────────────────────────────────────
     *
     * Adding a service bdapps publishes later:
     *
     *   1. Add its URL variable to .env.example and to EndpointVariables in BdappsOptions
     *   2. Add one wrapper here that calls PostAsync with the new key.
     *
     * It inherits credential injection, the timeout, error mapping and the not-provisioned
     * guard for free. Do not build a parallel client.
     */
}

/// <summary>A non-S1000 application-level response.</summary>
public sealed class BdappsException : Exception
{
    public BdappsException(string statusCode, string statusDetail, string service)
        : base($"[{statusCode}] {statusDetail} ({service})")
    {
        StatusCode = statusCode;
        StatusDetail = statusDetail;
        Service = service;
    }

    public string StatusCode { get; }

    public string StatusDetail { get; }

    public string Service { get; }

    public bool IsRetryable => BdappsClient.Transient.Contains(StatusCode);

    public bool IsConfiguration => BdappsClient.Configuration.Contains(StatusCode);
}

File: templates/java/BdappsClient.java
package com.example.bdapps;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.math.BigDecimal;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * bdapps API client — Java port of templates/typescript/bdapps-client.ts.
 *
 * <p>One {@link #post} helper injects credentials, applies a timeout, and turns non-S1000
 * responses into typed errors. Every service is a thin wrapper that resolves its endpoint
 * through {@link BdappsConfig#requireEndpoint} — so calling an API your application was not
 * provisioned for fails locally with a clear message, rather than as E1309 from the platform.
 *
 * <p>Java 17 ({@code java.net.http.HttpClient}, {@code Stream.toList()}, switch expressions) and
 * Jackson — the Spring Boot 3 baseline. On Java 11 replace {@code .toList()} with
 * {@code .collect(Collectors.toList())}. Register it as a singleton bean; it is thread-safe and
 * the underlying {@code HttpClient} pools connections.
 *
 * <p>SERVER-SIDE ONLY.
 */
public final class BdappsClient {

  /** A single outbound call should never hang. Protocol constant, not config. */
  private static final Duration TIMEOUT = Duration.ofSeconds(15);

  /** Platform-side. Worth retrying with backoff. */
  private static final Set<String> TRANSIENT =
      Set.of(
          "E1316", "E1318", "E1319", "E1332", "E1341", "E1360", "E1363", "E1364", "E1600",
          "E1601", "E1602", "E1603");

  /** Provisioning or credentials are wrong. Retrying will never help. */
  private static final Set<String> CONFIGURATION =
      Set.of(
          "E1301", "E1302", "E1303", "E1304", "E1305", "E1306", "E1307", "E1309", "E1310",
          "E1311", "E1313", "E1315", "E1322", "E1323", "E1324", "E1327", "E1328", "E1329",
          "E1336", "E1371", "E1381", "E1383", "E1387");

  /** Codes meaning "the state you wanted already holds". */
  public static final String BENIGN_REGISTER = "E1351"; // user already registered
  public static final String BENIGN_UNREGISTER = "E1356"; // user not registered
  public static final String BENIGN_DEBIT = "E1379"; // transaction already completed

  /**
   * Mandatory on every CaaS call, with exactly one accepted value. It only ever varies by being
   * forgotten, so it is a constant here rather than a caller argument.
   */
  public static final String PAYMENT_INSTRUMENT = "MobileAccount";

  private final BdappsConfig config;
  private final ObjectMapper mapper = new ObjectMapper();
  private final HttpClient http;

  public BdappsClient(BdappsConfig config) {
    this.config = config;
    // bdapps hosts have served an incomplete certificate chain, which strict clients reject.
    // Do NOT install a trust-all TrustManager to work around it — that lets anyone on the path
    // read the applicationId and password that can charge your subscribers. Import the
    // intermediate CA into a truststore and point an SSLContext at it instead:
    //
    //   .sslContext(SSLContexts.custom().loadTrustMaterial(truststore, null).build())
    //
    // See references/09-security-best-practices.md.
    this.http = HttpClient.newBuilder().connectTimeout(TIMEOUT).build();
  }

  /* ── Errors ─────────────────────────────────────────────────────────────── */

  public static final class BdappsException extends RuntimeException {
    private final String statusCode;
    private final String statusDetail;
    private final String service;

    BdappsException(String statusCode, String statusDetail, String service) {
      super("[" + statusCode + "] " + statusDetail + " (" + service + ")");
      this.statusCode = statusCode;
      this.statusDetail = statusDetail;
      this.service = service;
    }

    public String statusCode() {
      return statusCode;
    }

    public String statusDetail() {
      return statusDetail;
    }

    public String service() {
      return service;
    }

    public boolean isRetryable() {
      return TRANSIENT.contains(statusCode);
    }

    public boolean isConfiguration() {
      return CONFIGURATION.contains(statusCode);
    }
  }

  /* ── Helpers ────────────────────────────────────────────────────────────── */

  /**
   * Normalise a subscriber address. The ONLY place {@code tel:} is added.
   *
   * <p>Accepts an already-prefixed address, a masked hash, +94…, 0094… or a local 07… number.
   */
  public static String toTelAddress(String msisdn) {
    String trimmed = msisdn == null ? "" : msisdn.trim();
    if (trimmed.isEmpty()) {
      throw new IllegalArgumentException("[bdapps] Empty subscriber address");
    }
    if (trimmed.toLowerCase(Locale.ROOT).startsWith("tel:")) {
      return trimmed;
    }
    String digits = trimmed.replaceAll("[\\s()-]", "").replaceFirst("^\\+", "");
    if (digits.startsWith("00")) {
      digits = digits.substring(2);
    }
    if (digits.startsWith("0") && digits.length() == 10) {
      digits = "94" + digits.substring(1);
    }
    return "tel:" + digits;
  }

  /** Mask a subscriber address for logging. Never log the raw value. */
  public static String maskAddress(String address) {
    String body = address == null ? "" : address.replaceFirst("(?i)^tel:", "");
    if (body.length() <= 6) {
      return "tel:***";
    }
    return "tel:"
        + body.substring(0, 3)
        + "*".repeat(body.length() - 6)
        + body.substring(body.length() - 3);
  }

  /** A unique, persistable idempotency key for a charge. Max 32 characters. */
  public static String generateExternalTrxId() {
    return UUID.randomUUID().toString().replace("-", "");
  }

  /* ── Core ───────────────────────────────────────────────────────────────── */

  private JsonNode post(String service, String url, Map<String, Object> body, String... benign) {
    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("applicationId", config.applicationId());
    payload.put("password", config.password());
    payload.putAll(body);

    JsonNode data;
    try {
      HttpRequest request =
          HttpRequest.newBuilder(URI.create(url))
              .timeout(TIMEOUT)
              .header("Content-Type", "application/json")
              .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(payload)))
              .build();
      HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
      // The HTTP status is deliberately not consulted: bdapps returns 200 for
      // application-level failures, and the real outcome is statusCode in the body.
      data = mapper.readTree(response.body());
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new IllegalStateException("[bdapps] " + service + " interrupted", e);
    } catch (Exception e) {
      throw new IllegalStateException("[bdapps] " + service + " transport failure", e);
    }

    String statusCode = data.path("statusCode").asText("");
    if ("S1000".equals(statusCode) || List.of(benign).contains(statusCode)) {
      return data;
    }
    throw new BdappsException(statusCode, data.path("statusDetail").asText(""), service);
  }

  /* ── SMS ────────────────────────────────────────────────────────────────── */

  /** Send an MT SMS to one or more subscribers. */
  public JsonNode sendSms(List<String> to, String message) {
    List<String> recipients = to.stream().map(BdappsClient::toTelAddress).toList();
    if (recipients.contains("tel:all")) {
      throw new IllegalArgumentException(
          "[bdapps] Use broadcastSms() for tel:all — broadcasts must be deliberate.");
    }
    return post(
        "sms-send",
        config.requireEndpoint("smsSend"),
        Map.of("message", message, "destinationAddresses", recipients));
  }

  public static final String BROADCAST_CONFIRMATION =
      "I_HAVE_VERIFIED_THIS_GOES_TO_ALL_SUBSCRIBERS";

  /**
   * Send to the ENTIRE subscribed base.
   *
   * <p>Deliberately separate from {@link #sendSms} so it can never be reached by accident —
   * check the subscriber base size first, and put an authorisation check in front of this.
   */
  public JsonNode broadcastSms(String message, String confirmation) {
    if (!BROADCAST_CONFIRMATION.equals(confirmation)) {
      throw new IllegalArgumentException("[bdapps] Broadcast confirmation token missing");
    }
    return post(
        "sms-send",
        config.requireEndpoint("smsSend"),
        Map.of("message", message, "destinationAddresses", List.of("tel:all")));
  }

  /* ── USSD ───────────────────────────────────────────────────────────────── */

  /**
   * Send a USSD screen.
   *
   * <p>{@code sessionId} MUST be the one the platform sent you. Use "mt-fin" for the final
   * screen — anything else leaves the session hanging until the network times out.
   */
  public JsonNode sendUssd(
      String sessionId, String destinationAddress, String message, String operation) {
    if (!Set.of("mt-init", "mt-cont", "mt-fin").contains(operation)) {
      throw new IllegalArgumentException("[bdapps] Invalid ussdOperation '" + operation + "'");
    }
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("message", message);
    body.put("sessionId", sessionId);
    body.put("ussdOperation", operation);
    body.put("destinationAddress", toTelAddress(destinationAddress));
    body.put("encoding", "440");
    body.put("version", "1.0");
    return post("ussd-send", config.requireEndpoint("ussdSend"), body);
  }

  /* ── Subscription ───────────────────────────────────────────────────────── */

  /**
   * Opt a subscriber in. Only call this with recorded, explicit consent.
   *
   * <p>E1351 (already registered) is accepted as success — the desired state holds.
   */
  public JsonNode register(String subscriberId) {
    return post(
        "subscription-register",
        config.requireEndpoint("subscriptionSend"),
        Map.of("subscriberId", toTelAddress(subscriberId), "action", "1", "version", "1.0"),
        BENIGN_REGISTER);
  }

  /**
   * Opt a subscriber out.
   *
   * <p>E1356 (not registered) is accepted as success — the desired state holds.
   */
  public JsonNode unregister(String subscriberId) {
    return post(
        "subscription-unregister",
        config.requireEndpoint("subscriptionSend"),
        Map.of("subscriberId", toTelAddress(subscriberId), "action", "0", "version", "1.0"),
        BENIGN_UNREGISTER);
  }

  /** Check one subscriber's status. For reconciliation, not per-request gating. */
  public JsonNode getSubscriptionStatus(String subscriberId) {
    return post(
        "subscription-status",
        config.requireEndpoint("subscriptionStatus"),
        Map.of("subscriberId", toTelAddress(subscriberId)));
  }

  /**
   * Subscriber base size. Needs no subscriber and charges nothing, which also makes it the best
   * connectivity and credential smoke test. {@code baseSize} comes back as a string.
   */
  public long queryBase() {
    JsonNode data =
        post("subscription-query-base", config.requireEndpoint("subscriptionQueryBase"), Map.of());
    return Long.parseLong(data.path("baseSize").asText("0"));
  }

  /* ── OTP ────────────────────────────────────────────────────────────────── */

  /**
   * Send an OTP to a plain mobile number.
   *
   * <p>Rate-limit per number AND per IP before calling, or the app becomes an SMS-bombing tool.
   * Keep the returned referenceNo server-side; never log it.
   */
  public JsonNode requestOtp(String subscriberId, Map<String, Object> applicationMetaData) {
    return post(
        "otp-request",
        config.requireEndpoint("otpRequest"),
        Map.of(
            "subscriberId", toTelAddress(subscriberId),
            "applicationMetaData", applicationMetaData));
  }

  /**
   * Verify an OTP. Valid 60 minutes, maximum 3 attempts — enforce those limits on your side too.
   * The returned subscriberId is the masked identifier to use for every subsequent call.
   */
  public JsonNode verifyOtp(String referenceNo, String otp) {
    return post(
        "otp-verify",
        config.requireEndpoint("otpVerify"),
        Map.of("referenceNo", referenceNo, "otp", otp));
  }

  /* ── CaaS ───────────────────────────────────────────────────────────────── */

  /**
   * Charge a subscriber's mobile account.
   *
   * <p>THIS MOVES REAL MONEY.
   *
   * <ul>
   *   <li>{@code externalTrxId} is your idempotency key. Generate it with {@link
   *       #generateExternalTrxId()}, PERSIST IT, then call this.
   *   <li>There are deliberately no retries here. A timeout does NOT mean the charge failed.
   *       Resolve unknown outcomes by re-calling with the SAME externalTrxId, or by reconciling
   *       — the platform de-duplicates on it. Never re-roll the id.
   *   <li>E1379 (already completed) is accepted as success.
   *   <li>Amount is {@link BigDecimal} — never {@code double}.
   * </ul>
   */
  public JsonNode debit(
      String subscriberId, BigDecimal amount, String externalTrxId, String currency) {
    if (externalTrxId == null || externalTrxId.isBlank()) {
      throw new IllegalArgumentException(
          "[bdapps] externalTrxId is required and must be persisted first");
    }
    if (externalTrxId.length() > 32) {
      throw new IllegalArgumentException("[bdapps] externalTrxId must be 32 characters or fewer");
    }
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("externalTrxId", externalTrxId);
    body.put("subscriberId", toTelAddress(subscriberId));
    body.put("paymentInstrumentName", PAYMENT_INSTRUMENT);
    body.put("amount", amount.toPlainString());
    body.put("currency", currency == null ? "BDT" : currency);
    return post("caas-direct-debit", config.requireEndpoint("caasDebit"), body, BENIGN_DEBIT);
  }

  /**
   * Query chargeable balance.
   *
   * <p>Advisory only: the balance can change between this and the debit. Always handle E1378 on
   * the debit regardless of what this returned.
   */
  public JsonNode queryBalance(String subscriberId, String currency) {
    return post(
        "caas-balance-query",
        config.requireEndpoint("caasBalance"),
        Map.of(
            "subscriberId", toTelAddress(subscriberId),
            "paymentInstrumentName", PAYMENT_INSTRUMENT,
            "currency", currency == null ? "BDT" : currency));
  }

  /* ── Extension point ────────────────────────────────────────────────────────
   *
   * Adding a service bdapps publishes later:
   *
   *   1. Add its URL variable to .env.example and to ENDPOINT_VARS in BdappsConfig
   *   2. Add one wrapper here:
   *
   *        public JsonNode newThing(Map<String, Object> input) {
   *          return post("new-thing", config.requireEndpoint("newThing"), input);
   *        }
   *
   * It inherits credential injection, the timeout, error mapping and the not-provisioned guard
   * for free. Do not build a parallel client.
   */
}
